common/flatpak-instance.c

```diff
@@ -641,6 +762,147 @@ flatpak_instance_new_for_id (const char *id)
   return flatpak_instance_new (dir);
 }
 
+/*
+ * flatpak_instance_claim_per_app_temp_directory:
+ * @app_id: $FLATPAK_ID
+ * @link_path: Path of a symbolic link to a subdirectory of @parent
+ * @parent: The directory in which we created the temporary directory,
+ *  such as /dev/shm or /tmp
+ * @locking_mode: F_RDLCK or F_WRLCK
+ * @lock_fd_out: (out) (not optional): Return a fd that holds a lock
+ *  in @path_out
+ * @path_out: (out) (not optional): Return the path to the directory
+ *  referenced by @link_path
+ *
+ * Try to take control of an existing per-app temporary directory
+ * referenced by @link_path, either for reuse or for deletion.
+ * Return %TRUE if we can.
+ *
+ * This is currently only used for /dev/shm, but it's designed to be
+ * equally usable for other non-user-owned directories like /tmp.
+ *
+ * We have to be careful here, because @link_path might be left over
+ * from a previous boot, and it probably points into a directory like
+ * /dev/shm or /tmp, where an attacker might recreate our directories,
+ * for example as symbolic links to somewhere they control. As a result,
+ * this function is security-sensitive, and needs to follow a policy of
+ * failing when an unexpected situation is detected.
+ *
+ * @error is not normally user-visible, and is mostly present to support
+ * debugging and unit testing.
+ *
+ * Returns: %TRUE if @link_path points to a suitable directory,
+ *  or %FALSE with @error set if it does not.
+ */
+gboolean
+flatpak_instance_claim_per_app_temp_directory (const char *app_id,
```

Comment:
As with this function's earlier existence in #4093, it's security-sensitive, so please review carefully.

I added test coverage for all the branches except `fstat()` failing (which I don't know how to trigger on-demand). To get full test coverage for it, you need to run tests as non-root, and pre-create `/tmp/flatpak-com.example.App-OwnedByRoot` owned by root (which I'm going to try to add to the CI).